### DISCUSSION TITLE
How board design is coming along.

### DISCUSSION THREAD

**Discussion by TheEccentricGenius** [KenWillmott/HD6303EVB#10] [General]
How has  the board been coming together?

I am still waiting to acquire my various parts, so no major board designing yet.

**Comment by KenWillmott** [comment]
The SAA1099 sound generator IC's arrived, so I am testing it now. Together with the TMS9918A VDP, it will make a game card. I am giving that more priority than the main board now because it is cool, and may attract more devs and generate new ideas. Mulling over, whether I should add a UART to it, it can share a clock source, and also support a USB keyboard/mouse via a CH9350 adapter board. That would make it a stand alone terminal.

**Reply by TheEccentricGenius** [reply]
That sounds very cool.

I'll just wait around a bit for my stuff to come and will get back to then if that's alright.

**Comment by TheEccentricGenius** [comment]
So my 6303RP has just arrived along with the 6350 serial chip....Haven't used them yet however as I still don't have the 7.???Mhz Crystal for timing.

How is things going down your end?
Have you finalized any of the game board or did you get sidetracked with the main board?

**Comment by KenWillmott** [comment]
The game board schematic is almost complete. I halted when I came to address decoding, considering the issues mentioned in the discussion thread about peripheral selects. When I reach some conclusion about that, it will be done fairly quickly. KiCad lacks footprints for the connectors, so I have to learn how to make those, which will take some time. But the groundwork is done, most of the circuit is already laid out. I don't expect any real problems in PCB layout due to the modest size, so it should be possible to complete within 2 weeks or so.

In the past few days, the thoughts about memory decoding have begun to coalesce. So after the game board is in fabrication, I will have a pretty clear idea how to attack the main board revision.

**Comment by TheEccentricGenius** [comment]
I think I'll use the HD6303RP with my custom OS I wrote a few years ago....And perhaps get the HD6303Y for your board.

**Comment by KenWillmott** [comment]
An HD6303RP design would be extremely welcome. I doubt that I'll have time for both, when it comes to the next processor board, it will likely be a 6809/6309. Possibly even a dual processor. We'll see.

By the way, to facilitate a common baud rate clock source, I am leaning towards adding a second serial port to the main board, rather than any peripheral board.

**Comment by TheEccentricGenius** [comment]
So I found some code that runs on a Arduino Uno...The code unlocks SDP(Software Data Protection) on the AT28Cxx chips.

I have given a link below for you....It is my port of the repo as I am changing it to fit my hardware.
[EEPROM Programmer](https://github.com/TheEccentricGenius/EccentricPROM/tree/main)

I supplied this info since you commented that you were having issues with using you AT28C64s, and thought this may be the reason.

Yes, I will definitely be doing a HD6303RP design as I've got the chip....I'll be putting the additional serial chip on my board too.
But however it might be a minimal design, in terms of only having one RAM chip....That isn't setting stone however.
I'll get the design to github as soon as I've actually started it in more depth....I've just almost completed the EEPROM programmer for it....
Oh, and still waiting for the 7.xxx Mhz crystals.
Possibly dumb question, but do you think 11.0592MHZ would work?

I think the hard bit will be designing the expansion header etc.

**Comment by KenWillmott** [comment]
With the 11.0592 crystal, the best baud rate you can get is 14,400 with TCON register = 5. The issue would be E clock, it's almost 3 MHz and a "B" part is only 2 MHz, IIRC. It also ties up Timer 2, if it's used for serial. Best option for the HD63B03R would be an 8MHz oscillator for a 2MHz E clock, and a 1.8432 oscillator connected as an external baud clock source, providing the fastest CPU clock and the fastest baud rate, 115,200. With SS2=0, the clock source is directly derived from E, timer 2 is available for other uses. It's hard to beat the serendipity of the 7.3728MHz crystal because, almost the best performance in both areas.

I will address the other issues in the next message.

**Comment by KenWillmott** [comment]
I never had to deal with EEPROM problems here, because it was possible to use my New Micros HC11 board for that. I did think, what about that, since most people don't have anything like that. Also didn't have to deal with locking, since the AT28C64 doesn't have any. Regards programmers, I'm less interested in designs, and more interested in just obtaining a PCB because of time constraints. I tried to find one online, they are rare and/or expensive. However at least, once the monitor is installed, it will be possible to do software firmware updates right on the MCU board, without even removing the EEPROM IC.

I have another path under consideration but it's a major detour - I wrote a 21 byte bootloader, for a ROM-less system, could be keyed in manually with switches or whatever, then a monitor could be loaded and take control to install itself. Currently under development.

Now, the expansion header shouldn't really give you too many problems, it's really not much more than an extension of the CPU bus signals. For the current M8 bus road map, the only unusual signal would be the BSEL* signal, but it's really just a decoded memory block somewhere in your memory map, and you don't even have to stick to any specific address range to be M8 compatible. M8 doesn't specify addresses at all, that is a consideration for some other standardization effort. Some other signals like MDIS* can be omitted (i.e. left N/C), provided that no peripheral cards that use it, are installed. You CAN expect that a peripheral like the game card, will not function without BSEL*.

What I expect to happen, is that the existing BSEL* and IOSEL* will not appear on the M8 bus, instead those will select on board devices, and there will be new decoding to provide a BSEL* that is exclusive to peripheral boards. I have not settled on precise address ranges for any of those yet. However the general principle is decided. One simple approach would be to assign the M8 BSEL* (on the next revision of the EVB-HD6303Y board) to the existing unused memory space at $C000.

**Reply by TheEccentricGenius** [reply]
How strange CSI(Catalyst/On Semi) incorporated it into their 28C64....I know because it's 8K because the read/write doesn't go past 8K but rolls back over again.

I may or not get a PCB design onto github depends on whether I want to tweak hardware to cover more chips or not.
I only mentioned the programmer because I perhaps incorrectly supposed you had problems with your EEPROM....Speaking of programmers that HC11 must of cost a bit.

So should I try to have the same memory layout as your latest EVB-HD6303Y board?
If so I would like to know what it will be for clarity(I know you've mentioned it a least once).

Wow....Pretty minimal bootloader.
Yes, I am going to be patient and wait a few more days for that crystal to come.

**Comment by KenWillmott** [comment]
I guess I got lucky with my 28C64's, none of two batches were locked. It was interesting looking at programmers, it turns out there is a PCB for the TommyPROM. The HC11 was free in the sense that I paid for it 20 years ago, the manufacturer was bought out and doesn't sell it any more.

In terms of software and the memory map, I plan to make any software that I create fairly configurable for memory layout, so it doesn't matter that much. I recommend as much contiguous RAM as possible, I aimed for 48k as you can see. If you want to see a quick summary, it's listed in the Wiki, and printed in a text box in the main board schematic.

As you know, the first page of memory is "special" on the 6800/6801, because of the direct mode addressing and lack of direct page register such as the 6809 has. If you look at the monitor source, you can see that I have already begun to allocate things there, like the stack and interrupt vector table. Those can both be relocated by a kernel, when that is available. However the page zero resource is precious, the monitor uses it so it can run without external RAM installed. I think, as software development proceeds, there should be some coordination of page zero use for that reason. To be honest, I haven't been able to document the software as fast as I have written it, and those issues will be addressed. For now, what it means, is that external RAM should also begin at $0000 so that the direct page is fully equipped with valid RAM. The only difference is 24 bytes, but that is about 1/10 of the page zero space, so it's still important.

The tiny bootloader can be tested, when I can figure out how to obtain pure binary output from the assembler. Because, that is what you have to provide the bootloader for input.

I will start a separate discussion thread for software.

**Reply by TheEccentricGenius** [reply]
I was positive that two of the CAT28C64's were SDP locked, yet I can't lock on unlock them anymore.....I must of been dreaming.

I think I will just stick with your HD63C03Y memory map it seems well thought out....

BTW what assembler are you using?
I've only got dasm...It can generate binary but not S-Records.

**Comment by KenWillmott** [comment]
The assembler is Alan Baldwin's ASxxxx and ASLink:
[https://shop-pdp.net/ashtml/asxxxx.php](url)
I've created a separate discussion for software...

**Reply by TheEccentricGenius** [reply]
Thanks for the link....Sorry if the question was in the wrong place.

I saw that website a while back but didn't download the assembler from it for some reason.

**Comment by KenWillmott** [comment]
No problem, one more thing before moving the software talk there... if you use that assembler, because it has a linker too, it's really worth creating a makefile, like this:
`../../bin/as6801 -l -o -s loader.asm`
`cat loader.lst`
`../../bin/aslink -s loader.rel`

**Comment by TheEccentricGenius** [comment]
I've finished the first revision of the HD6303RP schematic....
[HD6303RP repository](https://github.com/TheEccentricGenius/HD6303RP_EVB)
[HD6303RP discussions](https://github.com/TheEccentricGenius/HD6303RP_EVB/discussions/1)

**Comment by KenWillmott** [comment]
Nice. I opened a minor issue on the NAND circuit, an easy fix. I like the MCU mode you selected, it enables the on chip I/O port, at the same time providing "free" buffering for the address lines. Something I forgot to do, you might find it worthwhile, is to connect one of the I/O pins to NVBANK, so that the NVRAM can use both 16k banks under software control. Then you have the use of the entire 64k onboard RAM.

**Reply by TheEccentricGenius** [reply]
I'm a little confused about how to implement the NVBANK switching in software and the I/O pin.

The other problem is I would kind of like to keep 8 I/O pins, then you get a whole port free.

**Comment by KenWillmott** [comment]
You can use one of the mode configuration pins, since their input state is only read at boot time. Depending on the intended use, you can omit the recommended mux circuit for the 3 mode pins. In your case, all three have resistor pull up/downs on them so you can use P20 (Timer Input) or P22 (Serial Baud Clock Input), since they are not in use. NVBANK is an input, it won't affect the pin state at boot time.

For example. replace VCC on the NVBANK jumper header, with a connection to P20 (pin 8). Then, the lower 16k RAM is selected with the jumper to ground, and the mode circuit is not connected. With the jumper in the other position on the header, memory address A14 on the RAM is connected to P20. Then the lower bank is selected at boot, but if you configure P20 as an output I/O port later, you can write the select bit to it and be able to select either lower or upper bank of the RAM. The boot mode load resistor is a minimal load on the output port, it won't affect the operation of the output pin.

The only odd restriction is that you can't use P21 (pin 9) for I/O output, only input. So you can't use that pin for this purpose. If you find some external use for P20 later, you can just set the NVBANK jumper to ground, and there is no connection to P20 from the circuit.

It's the same with the 6303Y EVB, so I opened an issue on it here.

**Comment by KenWillmott** [comment]
The game board design raised another issue, E latency in the peripheral decode. Because E is the strobe signal for bus I/O, it is better to exclude it from any decode logic, until the last possible stage before any peripheral chip select. The existing design gates BSEL* and IOSEL* with E on the main board. That must change. In future, there will be only one bus select, BSEL*, and the three remaining decodes from U10A may be gated with E to provide on board selects for on board peripheral devices such as the 512k memory expansion address latch. The modified circuit uses an additional NAND gate, but it is already available at least with the 6303RP design, since it uses only three gates in the 74HC00 RW decode circuit.

Note, this is an important change to the M8 bus characteristics, all future designs using the bus must follow this pattern regardless of what processor is driving it.

Here is the modification to the decode circuit to accommodate that change:

![image](https://github.com/KenWillmott/HD6303EVB/assets/17345651/45d5f002-724e-4c80-b848-c0d8e2889765)

The two additional selects SEL0* and SEL1* can be used either as an additional bus select for another bus connector, or else used for an on board peripheral (and gated with E if the peripheral itself does not have a spare active high enable for E).

**Reply by TheEccentricGenius** [reply]
It has been awhile, what have you gotten up to now?

**Reply by KenWillmott** [reply]
I've designed a few more I/O boards, and video boards. Mainly, the 6303 interest branched into the TRS-80 MC-10 that uses an MC6803 CPU. I designed a 32k upgrade for that one. I have some branches that could be worked on, like a remake of the existing HD63C03Y board. I recently completed an HC11 board. Problem is, I've been narrowly focused on pushing out working prototypes and not so much on uploading them here, or writing software for them.

**Reply by TheEccentricGenius** [reply]
Sounds alright.
Sounds you are quite busy with the designing phase of things.

**Reply by KenWillmott** [reply]
68HC11 board will be hosted here very soon, a few days.

**Comment by TheEccentricGenius** [comment]
Did you ever end up getting the PCF8584 working?

**Reply by KenWillmott** [reply]
No, I didn't. However the HC11 project gave me an idea - maybe the SPI code can be back ported to 6303. That would mean an SD card disk and many other SPI peripherals could be added. It's easier since the SD code is already written and proven.

**Reply by TheEccentricGenius** [reply]
Okay, I will try to get them working on a breadboard with a microcontroller or dip switches.
I think SPI should be able to be ported to the 6303 project, there is a custom computer constructed out of logic gates out there with SPI....

**Reply by KenWillmott** [reply]
Be sure to use the SD adapter type that has 5V level translator on board

**Reply by TheEccentricGenius** [reply]
I have one homemade one with a resistor divider for a level translator....Perhaps I should consider buying a proper level translator board though.

**Reply by KenWillmott** [reply]
SPI is in slow mode, that will work going "down voltage". You can check the input threshold levels on the MCU you are using to see if it's proper going "up voltage", i.e. SD card output into MISO on the MCU. If those are NMOS/TTL levels then you're okay with the simple resistor interface on MOSI and direct in on MISO. If they are CMOS levels then it will work or not work depending on the phase of the moon. A CMOS MCU like the HC11 really needs the IC translator.

**Reply by TheEccentricGenius** [reply]
Thanks for the advice, it is much appreciated.

**Comment by TheEccentricGenius** [comment]
Hello, it is me....Bumping a rather old discussion thread(Sorry about that, my sincere apologies).

However I would like to thank you for inspiring me on the route of delving down into the details of how computers work and such a lot deeper.
In the past few months I bought two FPGA boards due to the fact I wanted a few breadboards...And designed plus developed a fully functional 8 bit cpu, it is by no means finished but is a huge achievement.

Recently I have gotten a little distracted on developing a basic Arduino phone that works a little better then the examples the internet presents.

I hope you have been in good health and are doing the things that you enjoy.

Cheers.
